components/sync/driver/brave_sync_service_impl_unittest.cc

```diff
@@ -270,4 +286,50 @@ TEST_F(BraveSyncServiceImplTest, ForcedSetDecryptionPassphrase) {
   OSCryptMocker::TearDown();
 }
 
+TEST_F(BraveSyncServiceImplTest, OnSelfDeviceInfoDeleted) {
+  OSCryptMocker::SetUp();
+  CreateSyncService(SyncServiceImpl::MANUAL_START);
+
+  brave_sync_service_impl()->Initialize();
+  EXPECT_FALSE(engine());
+
+  brave_sync_service_impl()->SetSyncCode(kValidSyncCode);
+  task_environment_.RunUntilIdle();
+
+  // Replace DataTypeManager with mock who gives CONFIGURING result on state()
+  // call. We need this to force SyncService::GetTransportState() give
+  // TransportState::CONFIGURING status to test behavior
+  // BraveSyncServiceImpl::OnSelfDeviceInfoDeleted
+  NiceMock<DataTypeManagerMock> data_type_manager_mock;
+  std::unique_ptr<DataTypeManager> data_type_manager_mock_ptr =
+      std::unique_ptr<DataTypeManager>(&data_type_manager_mock);
+  std::unique_ptr<DataTypeManager> data_type_manager_original =
+      std::move(brave_sync_service_impl()->data_type_manager_);
+  brave_sync_service_impl()->data_type_manager_ =
+      std::move(data_type_manager_mock_ptr);
+
+  ON_CALL(data_type_manager_mock, state())
+      .WillByDefault(Return(DataTypeManager::CONFIGURING));
+
+  EXPECT_EQ(brave_sync_service_impl()->GetTransportState(),
+            SyncServiceImpl::TransportState::CONFIGURING);
+
+  NiceMock<SyncServiceObserverMock> observer_mock;
+  brave_sync_service_impl()->AddObserver(&observer_mock);
+
+  // When OnSelfDeviceInfoDeleted arrived, but transport state is CONFIGURING,
+  // we must not stop and clear service.
+  EXPECT_CALL(observer_mock, OnStateChanged).Times(0);
+  brave_sync_service_impl()->OnSelfDeviceInfoDeleted(base::DoNothing());
+
+  brave_sync_service_impl()->RemoveObserver(&observer_mock);
+
+  // Revert DataTypeManager to its original pointer
+  brave_sync_service_impl()->data_type_manager_.release();
+  brave_sync_service_impl()->data_type_manager_ =
+      std::move(data_type_manager_original);
```

Comment:
Is this necessary? I don't see we need original data_type_manager after it in this test case.

Reply:
I made it to ensure we will have no problems at the destruction time.

Just tried to disable these lines - there are no side effects, so I will remove it.

Reply:
yes, thanks, fixed at https://github.com/brave/brave-core/pull/16117/commits/df1282272989107247b2021f1fd5c8b6fb09b095

Reply:
yes, thanks, fixed at https://github.com/brave/brave-core/pull/16117/commits/df1282272989107247b2021f1fd5c8b6fb09b095